stm32cube/stm32f4xx/CMakeLists.txt

```
@@ -24,7 +24,13 @@ zephyr_library_sources_ifdef(CONFIG_USE_STM32_HAL_DMA drivers/src/stm32f4xx_hal_
 zephyr_library_sources_ifdef(CONFIG_USE_STM32_HAL_DMA2D drivers/src/stm32f4xx_hal_dma2d.c)
 zephyr_library_sources_ifdef(CONFIG_USE_STM32_HAL_DMA_EX drivers/src/stm32f4xx_hal_dma_ex.c)
 zephyr_library_sources_ifdef(CONFIG_USE_STM32_HAL_DSI drivers/src/stm32f4xx_hal_dsi.c)
-zephyr_library_sources_ifdef(CONFIG_USE_STM32_HAL_ETH drivers/src/Legacy/stm32f4xx_hal_eth.c)
+if(CONFIG_ETH_STM32_HAL_API_V2)
+  zephyr_compile_definitions(-DHAL_ETH_MODULE_ENABLED)
+  zephyr_library_sources(drivers/src/stm32f4xx_hal_eth.c)
+elseif(CONFIG_USE_STM32_HAL_ETH)
+  zephyr_compile_definitions(-DHAL_ETH_LEGACY_MODULE_ENABLED)
+  zephyr_library_sources(drivers/src/Legacy/stm32f4xx_hal_eth.c)
+endif()
```

As mentioned above, If possible, I'd prefer this is done with minimal impact on the existing code, so this could be removed with as much little changes once we'll get rid of the legacy API:

```suggestion
if(CONFIG_ETH_STM32_HAL_API_V2)
  zephyr_library_sources(drivers/src/stm32f4xx_hal_eth.c)
else
  zephyr_library_sources(drivers/src/Legacy/stm32f4xx_hal_eth.c)
endif()
```